### PR TITLE
Add ./configure

### PR DESCRIPTION
**Pull request by karlkfi**
- Vendor semver_bash bash library
- rename common.mk -> make-common.mk (for grouping of make files)
- Remove complex auto-configuration from Makefile and common.mk
- Add auto configure mode for easy first time use
- Add interactive configure mode for saving persistent settings
- Add manual configuration mode (write make-config.mk) for CI

### CODE REVIEW

**Comment by karlkfi** [comment]
Example auto-config output with Docker for Mac:

```
$ ./configure --auto
Auto Mode Enabled

Enter DC/OS installer path [dcos_generate_config.sh]:
Entered (default): dcos_generate_config.sh
Installer not found.

Download the latest stable DC/OS installer [Y/n]?
Downloading installer...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  786M  100  786M    0     0  33.3M      0  0:00:23  0:00:23 --:--:-- 35.4M
Validating installer path...
Inspecting installer...
Unpacking installer...
Detecting DC/OS version...
Version: 1.9.1
Variant: oss

Detecting Docker server version...
Server Version: 17.06.0-ce

Detecting Docker kernel version...
Kernel Version: 4.9.31-moby

Detecting Docker for Mac memory...
Docker Memory: 6144

Detecting Docker storage driver...
Storage Driver: overlay2

Select inner Docker storage driver:
1) overlay
2) overlay2 (default)
3) aufs
4) devicemapper
5) [custom]
#?
Selected (default): overlay2

Select inner Docker version:
1) 1.11.2
2) 1.13.1 (default)
#?
Selected (default): 1.13.1

Detecting Docker host init system...
Init System: 'init'

Enable Mesos systemd support (requires systemd on host):
1) true
2) false (default)
#?
Selected (default): false

Select number of master nodes:
1) 1 (default)
2) 3
3) 5
4) 7
5) 9
6) [custom]
#?
Selected (default): 1

Enter number of agent nodes [1]:
Entered (default): 1

Enter number of public agent nodes [1]:
Entered (default): 1

Generating dcos-docker config...
DCOS_GENERATE_CONFIG_PATH := $(CURDIR)/dcos_generate_config.sh
DOCKER_VERSION := 1.13.1
DOCKER_STORAGEDRIVER := overlay2
MESOS_SYSTEMD_ENABLE_SUPPORT := false
MASTERS := 1
AGENTS := 1
PUBLIC_AGENTS := 1

Configuration Complete!
Run `make` to deploy DC/OS.
```

**Comment by karlkfi** [comment]
Example auto-config with Vagrant:

```
$ ./configure --auto
Auto Mode Enabled

Enter DC/OS installer path [dcos_generate_config.sh]:
Entered (default): dcos_generate_config.sh
Validating installer path...
Inspecting installer...
Unpacking installer...
Detecting DC/OS version...
Version: 1.9.1
Variant: oss

Detecting Docker server version...
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Server Version: 1.13.1

Detecting Docker kernel version...
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Kernel Version: 3.10.0-514.el7.x86_64

Detecting Docker storage driver...
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Storage Driver: overlay

Select inner Docker storage driver:
1) overlay (default)
2) overlay2
3) aufs
4) devicemapper
5) [custom]
#?
Selected (default): overlay

Select inner Docker version:
1) 1.11.2 (default)
2) 1.13.1
#?
Selected (default): 1.11.2

Detecting Docker host init system...
Init System: 'systemd'

Enable Mesos systemd support (requires systemd on host):
1) true (default)
2) false
#?
Selected (default): true

Select number of master nodes:
1) 1 (default)
2) 3
3) 5
4) 7
5) 9
6) [custom]
#?
Selected (default): 1

Enter number of agent nodes [1]:
Entered (default): 1

Enter number of public agent nodes [1]:
Entered (default): 1

Generating dcos-docker config...
DCOS_GENERATE_CONFIG_PATH := $(CURDIR)/dcos_generate_config.sh
DOCKER_VERSION := 1.11.2
DOCKER_STORAGEDRIVER := overlay
MESOS_SYSTEMD_ENABLE_SUPPORT := true
MASTERS := 1
AGENTS := 1
PUBLIC_AGENTS := 1

Configuration Complete!
Run `make` to deploy DC/OS.
```

**Comment by karlkfi** [comment]
@adamtheturtle Can you review this?

**Comment by adamtheturtle** [comment]
@karlkfi Yes I can, when I return from EuroPython (1 day delay minimum).

File: README.md [hunk 1]
```diff
@@ -91,6 +83,32 @@ Host Docker storage drivers of `overlay` and `aufs` are supported automatically.
...
+
+    **Manual Mode (example):**
+
+    ```console
+    cat > make-config.mk << EOM
+    MASTERS := 3
+    EOM
+    ```
+
+    See [make-defaults.mk](make-defaults.mk) for a full list of manually configurable options.
```
Comment (adamtheturtle) [inline README.md]: This is tricky but this isn't the full list of manually configurable options - in fact I regularly use `MASTER_MOUNTS` and the only reason for that existing is for configuration - it is empty by default.

However, every variable with `:=` is manually configurable.

**Comment by karlkfi** [comment]
@asridharan I'd love to have this whole project rewritten in Go, but it's an expensive undertaking and I'm only working on this part time. In the mean time, I've just been iterating on what we have in Make and bash.

File: README.md (hunk 1 above)
Comment (karlkfi) [inline README.md]: Yeah, I know how make works, I just wanted to pull out the ones that are likely to be overwritten by users and provide a way to do that more persistently. You can of course still overwrite them the traditional way, at runtime.

If you'd like MASTER_MOUNTS in `make-defaults.mk` I can move it there. Are there any others?

Comment (karlkfi) [inline README.md]: I refactored the volume variables to clean them up. Now there are volume mount variables for each node type:

```
# Custom Docker volume arguments (passed to docker run)
# Syntax: -v [<host-path>:]<container-path>
CUSTOM_MASTER_VOLUMES :=
CUSTOM_AGENT_VOLUMES :=
CUSTOM_PUBLIC_AGENT_VOLUMES :=
```

**Comment by karlkfi** [comment]
Reviews got held up. So i'm gonna just merge and we can refine later.

**Pull request merged**